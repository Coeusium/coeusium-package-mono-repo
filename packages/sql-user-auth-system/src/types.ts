export interface User {
  id: string;
  email: string;
  username: string;
  registration_date: Date;
  last_login: Date;
  verified: boolean;
  logged_in: boolean;
}

export interface UserPassword {
  user_id: string;
  password: string;
}

export interface UserLogin {
  user_id: string;
  login_date: Date;
}

export interface UserLogout {
  user_id: string;
  logout_date: Date;
}

export interface UserPasswordReset {
  user_id: string;
  token: string;
  date_requested: string;
  completed: string;
}

export interface AuthGroup {
  group_name: string;
  group_parent: string;
}

export interface AuthGroupMembership {
  user_id: string;
  group_name: string;
}
