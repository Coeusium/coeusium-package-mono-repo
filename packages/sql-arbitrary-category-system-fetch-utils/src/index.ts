export * from './sql-arbitrary-category-system-fetch-utils';
export type { Category, CategorySet } from './types';
