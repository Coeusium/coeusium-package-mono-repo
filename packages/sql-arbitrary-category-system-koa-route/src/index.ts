export * from './sql-arbitrary-category-system-koa-route';
export type { ConnectionInfo, Category, CategorySet } from './types';
