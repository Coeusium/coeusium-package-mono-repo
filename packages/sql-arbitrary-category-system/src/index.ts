export * from './sql-arbitrary-category-system';
export type { ConnectionInfo, Category, CategorySet } from './types';
