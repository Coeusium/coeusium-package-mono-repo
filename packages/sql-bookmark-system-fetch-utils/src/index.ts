export * from './sql-bookmark-system-fetch-utils';
export type { Bookmark } from './types';
