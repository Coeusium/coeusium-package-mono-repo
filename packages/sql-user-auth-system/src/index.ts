export * from './sql-user-auth-system';
export type {
  User,
  UserPassword,
  UserLogin,
  UserLogout,
  UserPasswordReset,
  AuthGroup,
  AuthGroupMembership,
} from './types';
