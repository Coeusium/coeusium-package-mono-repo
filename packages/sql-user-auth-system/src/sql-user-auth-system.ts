import {
  ConnectionInfo,
  execute,
  setupDatabase,
  idExistsInTable,
  getFromTableById,
  listFromTable,
  deleteFromTableById,
  readSQLFiles,
} from '@almaclaine/mysql-utils';
import {
  InvalidEmail,
  validateEmail,
  makeId,
  timeNowString,
} from '@almaclaine/general-utils';
import { User } from './types';

export async function setupUserAuthSystem(dbInfo: ConnectionInfo) {
  await setupDatabase(dbInfo, 'sql_user_auth_system', await readSQLFiles());
}

export async function destroy() {
  execute.destroyConnections();
}

export async function userIdExists(dbInfo: ConnectionInfo, id: string) {
  return await idExistsInTable(dbInfo, 'user', id);
}

export async function addUser(
  dbInfo: ConnectionInfo,
  email: string,
  username: string,
) {
  if (!validateEmail(email))
    throw new InvalidEmail('Invalid email address passed to addUser');
  const sql = `INSERT INTO user (id, email, username) VALUES (?, ?, ?);`;
  let id = makeId();
  while (await userIdExists(dbInfo, id)) id = makeId();
  await execute(dbInfo, sql, [id, email, username]);
  return id;
}

export async function getUserById(dbInfo: ConnectionInfo, id: string) {
  return await getFromTableById<User>(dbInfo, 'user', id);
}

export async function getUserByUsername(
  dbInfo: ConnectionInfo,
  username: string,
) {
  const sql = `SELECT * FROM user WHERE username = ? LIMIT 1`;
  return ((await execute(dbInfo, sql, [username]))[0] as User) || null;
}

export async function getUserByEmail(dbInfo: ConnectionInfo, email: string) {
  const sql = `SELECT * FROM user WHERE email = ? LIMIT 1`;
  return ((await execute(dbInfo, sql, [email]))[0] as User) || null;
}

export async function listUsers(dbInfo: ConnectionInfo, page = 0, limit = 20) {
  return await listFromTable<User>(dbInfo, 'user', page, limit);
}

export async function deleteUserById(dbInfo: ConnectionInfo, id: string) {
  await deleteFromTableById(dbInfo, 'user', id);
}

export async function deleteUserByUsername(
  dbInfo: ConnectionInfo,
  username: string,
) {
  const sql = `DELETE FROM user WHERE username = ?;`;
  await execute(dbInfo, sql, [username]);
}

export async function deleteUserByEmail(dbInfo: ConnectionInfo, email: string) {
  const sql = `DELETE FROM user WHERE email = ?;`;
  await execute(dbInfo, sql, [email]);
}

export async function updateUserEmail(
  dbInfo: ConnectionInfo,
  id: string,
  email: string,
) {
  const sql = `UPDATE user SET email = ? WHERE id = ?;`;
  await execute(dbInfo, sql, [email, id]);
}

export async function updateUserUsername(
  dbInfo: ConnectionInfo,
  id: string,
  username: string,
) {
  const sql = `UPDATE user SET username = ? WHERE id = ?;`;
  await execute(dbInfo, sql, [username, id]);
}

export async function updateUserLastLogin(dbInfo: ConnectionInfo, id: string) {
  const sql = `UPDATE user SET last_login = ? WHERE id = ?;`;
  const login_date = timeNowString();
  return await execute(dbInfo, sql, [login_date, id]);
}

// const dbInfo = {
//     host: 'localhost',
//     user: 'root',
//     database: 'sql_user_auth_system',
//     password: process.env.MYSQL_PW
// };
//
// (async () => {
//     await setupUserAuthSystem(dbInfo);
// })()

// export async function addTag(dbInfo: ConnectionInfo, tag: string) {
//     const sql = `INSERT INTO tag (id, tag) VALUES (?, ?);`;
//     let id = makeId();
//     while (await tagIdExists(dbInfo, id)) id = makeId();
//     await execute(dbInfo, sql, [id, tag]);
//     return id;
// }
//
// export async function getTag(dbInfo: ConnectionInfo, id: string) {
//     const sql = `SELECT * FROM tag WHERE id = ? LIMIT 1`;
//     return (await execute(dbInfo, sql, [id]))[0] as Tag || null;
// }
//
// export async function listTags(dbInfo: ConnectionInfo, page = 0, limit = 20) {
//     const sql = `SELECT * FROM tag LIMIT ? OFFSET ?`;
//     const offset = limit * page;
//     return (await execute(dbInfo, sql, [limit, offset])) as Tag[] || [];
// }
//
// export async function deleteTag(dbInfo: ConnectionInfo, id: string) {
//     const sql = `DELETE FROM tag WHERE id=?`;
//     await execute(dbInfo, sql, [id]);
// }
