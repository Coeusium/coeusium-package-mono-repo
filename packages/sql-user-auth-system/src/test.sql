DROP TRIGGER IF EXISTS noSelfParentInsertTrigger;
CREATE Trigger noSelfParentInsertTrigger BEFORE Insert ON auth_group
    FOR EACH ROW
    BEGIN
        IF(new.group_name = new.group_parent) THEN
            SIGNAL SQLSTATE '45000'
            SET MESSAGE_TEXT = 'group_name and group_parent cannot be the same'
        END IF;
    END;
