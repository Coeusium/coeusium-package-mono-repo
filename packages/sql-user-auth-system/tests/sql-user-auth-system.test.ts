import * as sqlUserAuthSystem from '../src';

describe('sql-user-auth-system-test', () => {
  it('Has Correct API', () => {
    const keys = Object.keys(sqlUserAuthSystem);
    expect(keys.length).toBe(14);
    expect(keys).toMatchSnapshot();
  });
});
