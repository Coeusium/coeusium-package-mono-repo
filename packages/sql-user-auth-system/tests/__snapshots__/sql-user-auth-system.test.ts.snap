// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sql-user-auth-system-test Has Correct API 1`] = `
Array [
  "setupUserAuthSystem",
  "destroy",
  "userIdExists",
  "addUser",
  "getUserById",
  "getUserByUsername",
  "getUserByEmail",
  "listUsers",
  "deleteUserById",
  "deleteUserByUsername",
  "deleteUserByEmail",
  "updateUserEmail",
  "updateUserUsername",
  "updateUserLastLogin",
]
`;
